p4include/v1model.p4

```diff
@@ -88,27 +88,123 @@ enum MeterType {
+     *              updated, normally a value in the range [0,
+     *              size-1].  If index >= size, no meter state will be
+     *              updated.
+     * @param result Type T must be bit<W> with W >= 2.  When index is
+     *              in range, the value of result will be assigned 0
+     *              for color GREEN, 1 for color YELLOW, and 2 for
+     *              color RED (see RFC 2697 and RFC 2698 for the
+     *              meaning of these colors).  When index is out of
+     *              range, the final value of result is not specified,
+     *              and should be ignored by the caller.
+     */
     void execute_meter<T>(in bit<32> index, out T result);
 }
 
 extern direct_meter<T> {
     direct_meter(MeterType type);
+    /***
+     * read() may only be called within an action executed as a result
+     * of matching a table entry, of a table that has a direct_meter
+     * associated with it.  Calling read() causes the meter state
+     * associated with the matched table entry to be read, modified,
+     * and written back, atomically relative to the processing of
+     * other packets, and an integer encoding of one of the colors
+     * green, yellow, or red to be written to the result out
+     * parameter.
+     *
+     * TBD: Is the meter state updated even if an action is executed
+     * that does not call the read() method?
+     *
+     * @param result Type T must be bit<W> with W >= 2.  The value of
+     *              result will be assigned 0 for color GREEN, 1 for
+     *              color YELLOW, and 2 for color RED (see RFC 2697
+     *              and RFC 2698 for the meaning of these colors).
+     */
     void read(out T result);
 }
 
 extern register<T> {
     register(bit<32> size);
+    /***
+     * read() reads the state of the register array stored at the
+     * specified index, and returns it as the value written to the
+     * result parameter.
+     *
+     * @param index The index of the register array element to be
+     *              read, normally a value in the range [0, size-1].
+     * @param result Only types T that are bit<W> are currently
+     *              supported.  When index is in range, the value of
+     *              result becomes the value read from the register
+     *              array element.  When index is out of range, the
+     *              final value of result is not specified, and should
+     *              be ignored by the caller.
+     */
     void read(out T result, in bit<32> index);
+    /***
+     * write() writes the state of the register array at the specified
+     * index, with the value provided by the value parameter.  If you
+     * wish to perform a read() followed later by a write() to the
+     * same register array element, and you wish the read-modify-write
+     * sequence to be atomic relative to other processed packets, you
+     * must execute them in a P4_16 block annotated with an @atomic
```

Comment:
In practice, bmv2 considers that all actions and parser states are `@atomic`. However, for bmv2 it is not possible at the moment to have control blocks be `@atomic`

Reply:
Can BMv2 have multiple threads performing ingress processing in parallel, for different subsets of input ports?

Or multiple threads performing egress processing in parallel, for different subsets of output ports?

If so, does it cause action execution to be atomic by using a lock every time an action is executed?

If not, how is action execution atomic?

Reply:
Each register has an associated mutex. When executing an action, we acquire the mutexes for all the registers referenced in the action, in a predetermined order.

Reply:
Interesting.  I've just pushed another commit that attempts to describe this atomicity guarantee provided by the BMv2 implementation.

Reply:
I wonder if that kind of details belong in v1model.p4 or in behavioral-model/docs. Based on the P4 spec, programmers should still be using `@atomic`, and the bmv2 compiler should complain if the scope of `@atomic` is too large. In the future, there is nothing preventing bmv2 from supporting arbitrary `@atomic` blocks.

Reply:
Yes, v1model can be supported by other implementations besides bmv2. We should not be too specific in this file. I actually don't know if there are other implementations, though.

Reply:
I am fine moving BMv2 specific parts of this documentation elsewhere if you prefer.  I do think the BMv2-specific support/limitations should be documented at least once somewhere.  I will make another commit that attempts to remove BMv2 specifics from these changes.

Reply:
I have added a commit to this PR that removes the parts of the documentation that I believe to be BMv2-specific.  Those parts have been put into a separate PR here: https://github.com/p4lang/behavioral-model/pull/741

Reply:
To be strictly accurate, any or all of what is included in this PR _might_ differ in another v1model implementation.  I do not have any knowledge of other v1model implementations to check, though.